client-rpc/src/rpc/websocket_core.rs

```diff
@@ -0,0 +1,382 @@
+            println!(
+                "drop block {} current={} max={}",
+                height, current, self.max_height
+            );
+            return;
+        }
+        println!("******************* {} {}", height, kind);
+        let m = serde_json::to_string(&value).unwrap();
+        let m2: Block = serde_json::from_str(&m).unwrap();
+
+        let _ = self.write_block(height, &m2);
+    }
+
+    // low level block processing
+    pub fn write_block(&self, block_height: u64, block: &Block) -> Result<()> {
+        let block_results = self.client.block_results(block_height)?;
+
+        let block_time = block.time();
+
+        let transaction_ids = block_results.transaction_ids()?;
+        let block_filter = block_results.block_filter()?;
+
+        let wallet = self.get_current_wallet();
+        let unencrypted_transactions = self.check_unencrypted_transactions(
+            &block_filter,
+            wallet.staking_addresses.as_slice(),
+            block,
+        )?;
+
+        let block_header = BlockHeader {
+            block_height,
+            block_time,
+            transaction_ids,
+            block_filter,
+            unencrypted_transactions,
+        };
+
+        self.block_handler
+            .on_next(block_header, &wallet.view_key, &wallet.private_key)?;
+        Ok(())
+    }
+
+    // now one session is complete
+    // let's wait some time to next blocks
+    pub fn change_to_wait(&mut self) {
+        self.state = WebsocketState::WaitProcess;
+        self.state_time = SystemTime::now();
+    }
+
+    // tx channel for this thread
+    pub fn get_queue(&self) -> Sender<OwnedMessage> {
+        self.my_sender.clone()
+    }
+
+    // because everything is done via channel
+    // no mutex is necessary
+    // wallet can be added in runtime
+    pub fn add_wallet(&mut self, name: String, passphrase2: String) -> JsonResult<()> {
+        let passphrase: SecUtf8 = passphrase2.into();
+        println!("add_wallet ***** {} {}", name, passphrase);
+        let view_key = self
+            .wallet_client
+            .view_key(&name, &passphrase)
+            .map_err(to_rpc_error)?;
+        let private_key = self
+            .wallet_client
+            .private_key(&passphrase, &view_key)
+            .map_err(to_rpc_error)?
+            .ok_or_else(|| Error::from(ErrorKind::WalletNotFound))
+            .map_err(to_rpc_error)?;
+
+        let staking_addresses = self
+            .wallet_client
+            .staking_addresses(&name, &passphrase)
+            .map_err(to_rpc_error)?;
+
+        let info = WalletInfo {
+            name: name.to_string(),
+            staking_addresses,
+            view_key: view_key.clone(),
+            private_key,
+        };
+
+        self.wallets.push(info.clone());
+        println!("wallets length {}", self.wallets.len());
+        Ok(())
+    }
+
+    // Value is given from websocket_rpc
+    // received
+    fn do_parse(&mut self, value: Value) -> Option<()> {
```

Comment:
Instead of this manual parsing, it'll be easier to define the datatypes that capture the information received from RPC, use serde deserializer auto-derivation and just use `serde_json::...` -- just like it's done elsewhere in client-rpc, e.g.: https://github.com/crypto-com/chain/blob/156bd0299f76b3567016079b82a4860324f0a45b/client-rpc/src/server.rs#L167

Reply:
i agree, too.
but used this for inter thread communiation.

api cannot be called directly in other threads.